src/backend/distributed/executor/multi_utility.c

```diff
@@ -883,6 +897,228 @@ ProcessAlterObjectSchemaStmt(AlterObjectSchemaStmt *alterObjectSchemaStmt,
+
+	SavedMultiShardCommitProtocol = MultiShardCommitProtocol;
+	MultiShardCommitProtocol = COMMIT_PROTOCOL_BARE;
+	ExecuteModifyTasksWithoutResults(taskList);
+}
+
+
+/*
+ * IsSupportedDistributedVacuumStmt returns whether distributed execution of a
+ * given VacuumStmt is supported. The provided relationId (if valid) represents
+ * the table targeted by the provided statement.
+ *
+ * Returns true if the statement requires distributed execution and returns
+ * false otherwise; however, this function will raise errors if the provided
+ * statement needs distributed execution but contains unsupported options.
+ */
+static bool
+IsSupportedDistributedVacuumStmt(Oid relationId, VacuumStmt *vacuumStmt)
+{
+	const char *stmtName = (vacuumStmt->options & VACOPT_VACUUM) ? "VACUUM" : "ANALYZE";
+
+	if (vacuumStmt->relation == NULL)
+	{
+		/* WARN and exit early for local unqualified VACUUM commands */
+		ereport(WARNING, (errmsg("not propagating %s command to worker nodes", stmtName),
+						  errhint("Provide a specific table in order to %s "
+								  "distributed tables.", stmtName)));
+
+		return false;
+	}
+
+	if (!OidIsValid(relationId) || !IsDistributedTable(relationId))
+	{
+		return false;
+	}
+
+	if (!EnableDDLPropagation)
+	{
+		/* WARN and exit early if DDL propagation is not enabled */
+		ereport(WARNING, (errmsg("not propagating %s command to worker nodes", stmtName),
+						  errhint("Set citus.enable_ddl_propagation to true in order to "
+								  "send targeted %s commands to worker nodes.",
+								  stmtName)));
+	}
+
+	if (vacuumStmt->options & VACOPT_VERBOSE)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("the VERBOSE option is currently unsupported in "
+							   "distributed %s commands", stmtName)));
+	}
+
+	if (vacuumStmt->va_cols != NIL)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("specifying a column list is currently unsupported in "
+							   "distributed %s commands", stmtName)));
+	}
+
+	return true;
+}
+
+
+/*
+ * VacuumTaskList returns a list of tasks to be executed as part of processing
+ * a VacuumStmt which targets a distributed relation.
+ */
+static List *
+VacuumTaskList(Oid relationId, VacuumStmt *vacuumStmt)
+{
+	List *taskList = NIL;
+	List *shardIntervalList = NIL;
+	ListCell *shardIntervalCell = NULL;
+	uint64 jobId = INVALID_JOB_ID;
+	int taskId = 1;
+	StringInfo vacuumString = DeparseVacuumStmtPrefix(vacuumStmt);
+	const int vacuumPrefixLen = vacuumString->len;
+	Oid schemaId = get_rel_namespace(relationId);
+	char *schemaName = get_namespace_name(schemaId);
+	char *tableName = get_rel_name(relationId);
+
+	/* lock relation metadata before getting shard list */
+	LockRelationDistributionMetadata(relationId, ShareLock);
+
+	shardIntervalList = LoadShardIntervalList(relationId);
+
+	/* grab shard lock before getting placement list; ensure one-at-a-time VACUUM */
+	LockShardListMetadata(shardIntervalList, ExclusiveLock);
```

Comment:
I was about to say that that's true, and that we don't need the lock.   Just to remember that that's unfortunately not true. VACUUM is special™. It starts its own transaction / locks:
```
void
vacuum(int options, RangeVar *relation, Oid relid, VacuumParams *params,
	   List *va_cols, BufferAccessStrategy bstrategy, bool isTopLevel)
{
...
	 * Decide whether we need to start/commit our own transactions.
	 *
	 * For VACUUM (with or without ANALYZE): always do so, so that we can
	 * release locks as soon as possible.  (We could possibly use the outer
	 * transaction for a one-table VACUUM, but handling TOAST tables would be
	 * problematic.)
	 *
...
	if (options & VACOPT_VACUUM)
		use_own_xacts = true;
...
	if (use_own_xacts)
	{
...
		/* matches the StartTransaction in PostgresMain() */
		CommitTransactionCommand();
...
	/*
	 * Finish up processing.
	 */
	if (use_own_xacts)
	{
		/*
		 * This matches the CommitTransaction waiting for us in
		 * PostgresMain().
		 */
		StartTransactionCommand();
...
```

So we can't rely on vacuum locks, like at all.